### PR TITLE
Respect Sas token in conn string

### LINKED ISSUE #8859: Azure Blob Storage: Connection String SAS
Trying to upload blob to Azure Blob Storage using SAS based connection string. Crashed with the exception below. When using Access Key based connection string all works fine.

The specified resource does not exist.
RequestId:4affd386-201e-00d6-02f2-9f6b28000000
Time:2019-11-20T22:32:27.5499019Z
Status: 404 (The specified resource does not exist.)

ErrorCode: ResourceNotFound

Headers:
Server: Windows-Azure-Blob/1.0,Microsoft-HTTPAPI/2.0
x-ms-request-id: 4affd386-201e-00d6-02f2-9f6b28000000
x-ms-client-request-id: a73be4bc-348c-480d-8474-45ef2c806a57
x-ms-version: 2019-02-02
x-ms-error-code: ResourceNotFound
Date: Wed, 20 Nov 2019 22:32:26 GMT
Content-Length: 223
Content-Type: application/xml

Connecting string is copied from Azure portal, looks smth like this:
"BlobEndpoint=https://xxx.blob.core.windows.net/;SharedAccessSignature=sv=2019-02-02&ss=b&srt=sco&sp=rwdlac&se=2029-11-21T04:57:39Z&st=2019-11-19T20:57:39Z&spr=https&sig=<sig>"

***Exception or Stack Trace***
 at Azure.Storage.Blobs.BlobRestClient.BlockBlob.UploadAsync_CreateResponse(Response response)
   at Azure.Storage.Blobs.BlobRestClient.BlockBlob.<UploadAsync>d__0.MoveNext()

***Code Snippet***
client.GetBlobClient(blob.BlobName).UploadAsync(stream, allowOverwrite, cancellationToken);

 - Version of the Library used: 12.0.0.0

### PR DESCRIPTION
**Pull request by JoshLove-msft**
- Update StorageConnectionString to include Sas token in the various Endpoint properties. 
- Remove ctors from StorageConnectionString that are only used in tests, and instead add corresponding factory methods to TestExtensions
- Remove Table related code in StorageConnectionString and tests as this SDK doesn't support table endpoints.

Fixes #8859

### CODE REVIEW

File: sdk/storage/Azure.Storage.Common/src/Shared/SasExtensions.cs
Comment (JoshLove-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/SasExtensions.cs]: Moved this down for consistency with the ordering the Portal uses when constructing the Sas signature.

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs
Comment (JoshLove-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs]: Allowed this to be settable for testing.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs
Comment (jongio) [inline sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: Technically when discussing URIs, this part is called the scheme, not protocol.

Comment (jongio) [inline sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: Please consider using `Uri.UriSchemeHttps` and `Uri.UriSchemeHttp` instead of hard coding https and http.

**Review by jongio — CHANGES_REQUESTED** [review]
Thank you Josh!

File: sdk/storage/Azure.Storage.Blobs/tests/SessionRecords/BlobBaseClientTests/Ctor_ConnectionString_Sas.json
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/SessionRecords/BlobBaseClientTests/Ctor_ConnectionString_Sas.json]: These test recordings are used when running the tests in Playback mode which is what happens during CI. It may make sense to sanitize out the account name, but it isn't really required. If we made a change like that, we would have to re-record all of our tests.
This [link ](https://github.com/Azure/azure-sdk-for-net/blob/ab2e85209e9d665b58653f4dd13b6c11eaa5bc0a/Documentation/Track2TestFramework.md)has more info about the test infrastructure.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: As an FYI, this and the bodies of the other new factory methods were extracted as-is from StorageConnectionString constructors, which were only used by tests. 

The parameter is object, because it can be either a SharedAccessSignatureCredentials or a StorageSharedKeyCredential. We probably didn't have them inherit from a common type so as to avoid adding types that customers wouldn't actually need to use.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: Sure

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: Updated

Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: Actually, we have these defined already within Constants so I can use those.

Comment (jongio) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: Can you look into changing the constants to use the Uri.UriSchemeHttp(s) properties?  This is a nit, but if it's already there in a const as part of .NET, then we should use it.

**Review by seanmcc-msft — APPROVED** [review]
Looks good.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: I think this is getting a little outside the scope of this PR.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs]: Blobs and All should be sufficient for this file.  Adding additional tests in the Files and Queue test projects might be a good idea.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: I will hold off on this. One consideration is that we try to group all constants in one file for discoverability. I can't redirect to the Uri property either because our field is a "const".

File: sdk/storage/Azure.Storage.Common/src/Shared/SasExtensions.cs
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/SasExtensions.cs]: Mind adding a comment in the source explaining why the order's important?

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs]: Do we know for sure the endpoint will never have any query params?  Especially since connection strings aren't well defined anywhere?

File: sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs]: Added a Read-only

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs]: Nit - is it worth sharing a single builder and just changing the host between Uri creation?

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: Azure.Core didn't want to add a base `Credential` type so we're stuck with `object`.

**Review by tg-msft — APPROVED** [review]
Please get escrow approval before merging this change.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs]: And a Write, Delete

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs]: Would you recommend using HttpUtility.ParseQueryString to append sasToken?

Comment (tg-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs]: No, I don't think we need to parse it - I just don't want to stomp any values.  I'd create the builder first without setting `Query`.  If `sasToken` isn't null/empty, then we can check whether `builder.Query` has a value before appending (and possibly squeezing in a `&`).

**Comment by tg-msft** [comment]
Also - do we need to verify HTTPS if a SAS is used via connection string now?

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs]: Updated

**Comment by JoshLove-msft** [comment]
> Also - do we need to verify HTTPS if a SAS is used via connection string now?

We should add this for SAS in general, not just for connection string, but let's do it in a separate PR.

File: sdk/storage/Azure.Storage.Common/src/Shared/SasExtensions.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/SasExtensions.cs]: Added

**Comment by JoshLove-msft** [comment]
> > Also - do we need to verify HTTPS if a SAS is used via connection string now?
> 
> We should add this for SAS in general, not just for connection string, but let's do it in a separate PR.

Ended up adding this in here.

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageConnectionString.cs]: ?

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/TestExtensions.cs]: Nit - use `string.CompareOrdinal` or drop the `System.` prefix?

**Review by tg-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs]: Added

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareServiceClient.cs
Comment (JoshLove-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareServiceClient.cs]: This ctor is unused.

File: sdk/storage/Azure.Storage.Queues/src/QueueServiceClient.cs
Comment (JoshLove-msft) [inline sdk/storage/Azure.Storage.Queues/src/QueueServiceClient.cs]: Unused

**Comment by JoshLove-msft** [comment]
> > > Also - do we need to verify HTTPS if a SAS is used via connection string now?
> > 
> > 
> > We should add this for SAS in general, not just for connection string, but let's do it in a separate PR.
> 
> Ended up adding this in here.

Backing this out. This needs more discussion.

**Pull request merged**
